examples/AspNetCore/Controllers/WeatherForecastController.cs

```diff
@@ -19,8 +20,14 @@ public class WeatherForecastController : ControllerBase
         private static HttpClient httpClient = new HttpClient();
 
         [HttpGet]
-        public IEnumerable<WeatherForecast> Get()
+        public IEnumerable<WeatherForecast> Get(int? zipPostalCode = null)
         {
+            using var scope = EnrichmentScope.Begin(a =>
```

Comment:
This demos usage. Here the user wants to add "user.zip_postal" tag, if provided, on the HTTP span that could be created (enrichment only happens if the activity is created & sampled).

Reply:
Why don't we let users do Activity.Current?.AddTag(AddTag("user.zip_postal", zipPostalCode.Value.ToString())

Reply:
@cijothomas That's what this is doing 😄

* If the user does `Activity.Current?.AddTag("user.zip_postal", zipPostalCode.Value.ToString())` the data shows up on the currently running span. In the case of this code, that will be the http-in span.

* If the user does `EnrichmentScope.Begin` the data shows up on the NEXT span, aka child span, which does not exist yet when the code is executing. `EnrichmentScope` is basically allowing you to hang on to that contextual information until the Activity is available. In the case of this code, that will be the http-out span.

Reply:
I do like the concept. I've had customers want to query their data before and be able to group it by a combination of attributes on an automatically created span (like the http-out span) and attributes from one of their custom spans, though depending on what your backend is capable of this may not be easily possible if the attributes are spread across two spans.

This does mean that _all_ the spans created within the scope get "enriched" right? So, for instance in the case of gRPC which uses HTTP under the covers, both the gRPC span and HTTP span would get the `user.zip_postal` attribute. Well.. then again, I suppose the `EnrichmentAction` could be arbitrarily complex and you could make sure you target the correct span.  Maybe that's ok, but it may get a bit complicated since writing the `EnrichmentAction` depends on the user knowing what the instrumentation is doing.

All that said, yes I do like this, and while sure it could sometimes become complicated or lead to unintuitive results, I think there are many use cases this would benefit greatly.

Reply:
> This does mean that _all_ the spans created within the scope get "enriched" right?

No! The next Activity/Span created+sampled is enriched, then the scope is automatically closed. It does not flow to children. That is key. It's not meant to be baggage or correlation, there are already mechanisms/designs for that.

> As the recommended way to enrich current activity at any point in user code is to retrieve the current active span and enrich it, by doing tracer.CurrentSpan.SetAttribute or in .NET, Activity.Current?.SetAttribute.

Of course, agree 100%. But not always possible! EnrichmentScope is enabling a previously impossible scenario. Let me try to explain the user story for this.

I have some groups in my company working with OTel. I had the same ask from two different teams. Both expose an internal API that fronts a big database. Their spans are primarily HTTP-IN and SQL-OUT. Almost immediately both teams asked to enrich their SQL spans. The data (tags) added by the Sql Instrumentation are basic, taken off of the SqlConnection & SqlCommand. The code calling Sql has a bunch of _contextual_ information about what the library/API is doing. It knows the user, the area, the application, etc. That information is never passed on the SqlConnection or SqlCommand. The ask from these users was, how do we add this _contextual_ information on our SQL spans? Not possible. Both teams opted to turn off the Sql instrumentation and manually instrumented their code, adding their own data and re-doing what the instrumentation was doing.

The _contextual_ information is the key. You can't `Activity.Current.AddTag` that stuff because the Activity is created later, without the information around.

> The original issue #73 is asking for a way to enrich the current span with arbitrary information available from the httprequestmessage etc.

Yes. That could be done with options. You could get HttpRequest & HttpResponse, but not the contextual stuff. It also wouldn't help you with SQL. We could add SQL options, but you would just get SqlCommand & SqlConnection, which has limited utility.

What I'm doing here is a bit bigger than the ask on the issue. This solve is global, applies to any instrumentation the user might use, and solves the context thing, which was previously impossible.

But what if the user wanted the raw HttpRequest or raw HttpResponse? We could still have the options thing. I think all instrumentation should have basic settings for filters & enrichment. That doesn't undo the need for an enrichment scope thing though, because it's solving the _contextual_ information needing to be pushed on the next span.

By the way, using this enrichment pattern you could actually also solve the raw Http case. Like this:
```csharp
            using var scope = EnrichmentScope.Begin(a =>
            {
                if (zipPostalCode.HasValue)
                    a.AddTag("user.zip_postal", zipPostalCode.Value.ToString());

                HttpWebRequest request = (HttpWebRequest)a.GetCustomProperty("HttpWebRequest.Request");
                a.AddTag("http.content_type", request.ContentType);
            });
```

Reply:
Nicely put. The use case you describe is exactly what I've come up against in the past (and did not describe nearly as well 😄). This arbitrary contextual information is beneficial with the customers I've encountered because New Relic provides a query interface for querying span data. With this it would be nice to be able to query `SELECT * FROM Span WHERE http.url LIKE "%blah.com%" AND zipCode = 90210`. This query isn't possible if the zip code attribute is not on the HTTP span.

Though I'm still thinking that this is a hard problem. It seems all too easy to make unintentional changes. Like what would happen if the code you have was refactored to:

```
            using var scope = EnrichmentScope.Begin(a =>
            {
                // The span generated downstream from this call would get decorated?
                NewMethodCallHereThatDoesSomeWorkWithADatabaseAndCreatesASpan();

                if (zipPostalCode.HasValue)
                    a.AddTag("user.zip_postal", zipPostalCode.Value.ToString());

                HttpWebRequest request = (HttpWebRequest)a.GetCustomProperty("HttpWebRequest.Request");
                a.AddTag("http.content_type", request.ContentType);
            });
```

Reply:
@alanwest Yes in that case the scope would end up wrong. Not something I am torn up about necessarily. I just feel like... _why_ would you make a call right there? Is it an edge case? We could actually (probably) support it. If we close the scope(s) before calling the callback(s).

> Another idea I've pondered in the past is the idea of a scope provided with a callback where the callback is invoked when spans are created allowing you to be smart about what it's enriching... haven't thought this through too deeply. The context you want to add may not be available in the scope of the callback...

Can you go into that a bit more? Kind of sounds like what I'm doing here.

Reply:
> Not something I am torn up about necessarily. I just feel like... why would you make a call right there? Is it an edge case?

I'm not too concerned about it either, just raising it. Whether it is an edge case or not is hard to tell. Maybe a slightly more interesting example would be:

```csharp
            using var scope = EnrichmentScope.Begin(a =>
            {
                ...
            });

            // Maybe this method is owned by some other team or library out of my control.
            MethodThatMakesHttpOrSqlCallDownstreamThatIWantEnriched();
```

It's even possible that this code works one day, but then later the implementation changes and creates a span somewhere downstream.

I'm not suggesting we need to solve all the problems. I'm fine with limiting the situations we can feasibly support, I'm just thinking things through. I am 👍 on this idea and would be pretty interested in learning of any rough edges through people's experience.

> We could actually (probably) support it.

This does seem good if it can work.

---

> Can you go into that a bit more? Kind of sounds like what I'm doing here.

Ha! Sorry, I generated a jumbled 🕸️ of thought here. What was more on my mind was _not_ a scope, but maybe a delegate that HTTP instrumentation could be configured with that gets invoked on all HTTP calls enabling you to enrich at that point in time. But this idea definitely precludes you from accessing the important context you'd want to enrich with.

Reply:
> A delegate that HTTP instrumentation could be configured with that gets invoked on all HTTP calls enabling you to enrich at that point in time.

I'll do that too next. I'm all for adding points in the process users can hook into. IMO all the instrumentation should have standard delegates (more or less) for filters and enrichment. BUT, they are of limited utility. A delegate on Http options will get you HttpRequest, HttpResponse, and maybe Exception. There's only so much data on those that will be of interest. Something like our .NET Framework SqlClient, you only get a few strings. That's why I went for the enrichment scope, so users can add whatever they want from a closure. More power 🏭

Reply:
A thought stemming from one of my earlier questions: "This does mean that all the spans created within the scope get "enriched" right?"

What if this were an option? Something like:

```csharp
enum EnrichmentScopeOptions
{
    FirstSpan, // default behavior - only enrich the next span
    AllSpans   // enrich all spans within scope
}
```

Reply:
or TTL 😄

Reply:
Time to live?

Reply:
I just pushed new changes. Going to resolve this thread, please continue discussing though!